File: go/fetchers/kbs.go
package fetchers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/ducnhd/lotusmarket/go/types"
)

const kbsBaseURL = "https://kbbuddywts.kbsec.com.vn/sas/kbsv-stock-data-store/stock"

func KBS(ctx context.Context, ticker string) (*types.KBSQuote, error) {
	to := time.Now().Format("2006-01-02")
	from := time.Now().AddDate(0, 0, -7).Format("2006-01-02")
	url := fmt.Sprintf("%s/%s/historical-quotes?from=%s&to=%s", kbsBaseURL, ticker, from, to)
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("KBS request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("KBS HTTP %d", resp.StatusCode)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("KBS read body: %w", err)
	}
	var quotes []types.KBSQuote
	if err := json.Unmarshal(body, &quotes); err != nil {
		return nil, fmt.Errorf("KBS parse: %w", err)
	}
	if len(quotes) == 0 {
		return nil, fmt.Errorf("no KBS data for %s", ticker)
	}
	return &quotes[0], nil
}

File: python/lotusmarket/fetchers/kbs.py
"""KBS API — fundamental data (no API key needed)."""

from datetime import datetime, timedelta
from lotusmarket.types import KBSQuote

KBS_BASE_URL = "https://kbbuddywts.kbsec.com.vn/sas/kbsv-stock-data-store/stock"


def kbs(ticker: str) -> KBSQuote:
    """Fetch latest fundamentals (P/E, P/B, EPS, Beta, Dividend Yield)."""
    import httpx

    to_date = datetime.now().strftime("%Y-%m-%d")
    from_date = (datetime.now() - timedelta(days=7)).strftime("%Y-%m-%d")
    url = f"{KBS_BASE_URL}/{ticker}/historical-quotes?from={from_date}&to={to_date}"
    resp = httpx.get(url, timeout=15)
    resp.raise_for_status()
    quotes = resp.json()
    if not quotes:
        raise ValueError(f"No KBS data for {ticker}")
    q = quotes[0]
    return KBSQuote(
        date=q.get("TradingDate", ""),
        open=q.get("OpenPrice", 0),
        high=q.get("HighestPrice", 0),
        low=q.get("LowestPrice", 0),
        close=q.get("ClosePrice", 0),
        volume=q.get("TotalVol", 0),
        market_cap=q.get("MarketCapital", 0),
        pe=q.get("PE", 0),
        pb=q.get("PB", 0),
        eps=q.get("EPS", 0),
        bvps=q.get("BVPS", 0),
        beta=q.get("Beta", 0),
        dividend_yield=q.get("Yield", 0),
    )
